server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1345,15 +1346,20 @@ private void innerOpenEngineAndTranslog(final EngineConfig.OpenMode openMode, fi
         // we disable deletes since we allow for operations to be executed against the shard while recovering
         // but we need to make sure we don't loose deletes until we are done recovering
         config.setEnableGcDeletes(false);
-        if (openMode == EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG) {
-            // we have to set it before we open an engine and recover from the translog because
-            // acquiring a snapshot from the translog causes a sync which causes the global checkpoint to be pulled in,
-            // and an engine can be forced to close in ctor which also causes the global checkpoint to be pulled in.
-            final String translogUUID = store.readLastCommittedSegmentsInfo().getUserData().get(Translog.TRANSLOG_UUID_KEY);
-            final long globalCheckpoint = Translog.readGlobalCheckpoint(translogConfig.getTranslogPath(), translogUUID);
-            replicationTracker.updateGlobalCheckpointOnReplica(globalCheckpoint, "read from translog checkpoint");
-        }
-        createNewEngine(config);
+        try {
+            if (openMode == EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG) {
+                // we have to set it before we open an engine and recover from the translog because
+                // acquiring a snapshot from the translog causes a sync which causes the global checkpoint to be pulled in,
+                // and an engine can be forced to close in ctor which also causes the global checkpoint to be pulled in.
+                final String translogUUID = store.readLastCommittedSegmentsInfo().getUserData().get(Translog.TRANSLOG_UUID_KEY);
+                final long globalCheckpoint = Translog.readGlobalCheckpoint(translogConfig.getTranslogPath(), translogUUID);
+                replicationTracker.updateGlobalCheckpointOnReplica(globalCheckpoint, "read from translog checkpoint");
+            }
+            createNewEngine(config);
+        } catch (TranslogCorruptedException ex) {
+            store.markStoreCorrupted(new IOException("failed to open engine; translog corrupted", ex));
```

Comment:
I wonder if we shoudl move this to the engine constructor. We already have catch blocks there.